File: mouse_paint.py
#!/usr/bin/env python3
# mouse_paint.py - Рисовалка с мышкой на Python (Tkinter)
"""
Поддерживает: рисование кистью, выбор цвета, толщину, ластик, очистку, сохранение PNG.
"""
import tkinter as tk
from tkinter import colorchooser, filedialog, ttk
from PIL import Image, ImageDraw, ImageTk
import io

class MousePaint:
    def __init__(self, root):
        self.root = root
        self.root.title("🖌️ MousePaint - Python")
        self.root.geometry("900x700")
        self.root.resizable(True, True)

        self.color = "#000000"
        self.bg_color = "#ffffff"
        self.size = 5
        self.tool = "brush"  # brush, eraser
        self.last_x = None
        self.last_y = None
        self.undo_stack = []
        self.redo_stack = []
        self.max_undo = 20
        self.drawing = False

        # Холст
        self.canvas = tk.Canvas(self.root, bg=self.bg_color, cursor="cross")
        self.canvas.pack(fill=tk.BOTH, expand=True)
        self.canvas.bind("<Button-1>", self.start_draw)
        self.canvas.bind("<B1-Motion>", self.paint)
        self.canvas.bind("<ButtonRelease-1>", self.stop_draw)

        # Изображение для сохранения
        self.image = Image.new("RGB", (800, 600), self.bg_color)
        self.draw = ImageDraw.Draw(self.image)

        self.create_toolbar()
        self.bind_keys()

    def create_toolbar(self):
        toolbar = tk.Frame(self.root, bg="#f0f0f0", relief=tk.RAISED, bd=2)
        toolbar.pack(side=tk.TOP, fill=tk.X)

        tk.Button(toolbar, text="🖌️ Кисть", command=lambda: self.set_tool("brush")).pack(side=tk.LEFT, padx=2)
        tk.Button(toolbar, text="🧽 Ластик", command=lambda: self.set_tool("eraser")).pack(side=tk.LEFT, padx=2)

        tk.Label(toolbar, text="Цвет:").pack(side=tk.LEFT, padx=(10,0))
        self.color_btn = tk.Button(toolbar, bg=self.color, width=2, command=self.choose_color)
        self.color_btn.pack(side=tk.LEFT, padx=2)

        tk.Label(toolbar, text="Толщина:").pack(side=tk.LEFT, padx=(10,0))
        self.size_var = tk.IntVar(value=5)
        tk.Scale(toolbar, from_=1, to=20, orient=tk.HORIZONTAL, variable=self.size_var,
                 length=100, showvalue=0).pack(side=tk.LEFT)

        tk.Button(toolbar, text="↩️ Отменить", command=self.undo).pack(side=tk.LEFT, padx=5)
        tk.Button(toolbar, text="↪️ Повторить", command=self.redo).pack(side=tk.LEFT, padx=5)
        tk.Button(toolbar, text="🗑️ Очистить", command=self.clear).pack(side=tk.LEFT, padx=5)
        tk.Button(toolbar, text="💾 Сохранить", command=self.save).pack(side=tk.LEFT, padx=5)

    def set_tool(self, tool):
        self.tool = tool

    def choose_color(self):
        color = colorchooser.askcolor(color=self.color)[1]
        if color:
            self.color = color
            self.color_btn.config(bg=color)

    def start_draw(self, event):
        self.last_x = event.x
        self.last_y = event.y
        self.drawing = True
        self.push_undo()

    def paint(self, event):
        if not self.drawing:
            return
        x, y = event.x, event.y
        color = self.bg_color if self.tool == "eraser" else self.color
        size = self.size_var.get()
        self.canvas.create_line(self.last_x, self.last_y, x, y,
                                width=size, fill=color,
                                capstyle=tk.ROUND, smooth=True)
        self.draw.line([self.last_x, self.last_y, x, y], fill=color, width=size)
        self.last_x, self.last_y = x, y

    def stop_draw(self, event):
        self.drawing = False
        self.push_undo()

    def push_undo(self):
        # Сохраняем состояние в память
        with io.BytesIO() as output:
            self.image.save(output, format="PNG")
            data = output.getvalue()
        self.undo_stack.append(data)
        if len(self.undo_stack) > self.max_undo:
            self.undo_stack.pop(0)
        self.redo_stack.clear()

    def undo(self):
        if len(self.undo_stack) < 2:
            return
        self.redo_stack.append(self.undo_stack.pop())
        self.restore_from_data(self.undo_stack[-1])
        self.update_canvas()

    def redo(self):
        if not self.redo_stack:
            return
        data = self.redo_stack.pop()
        self.undo_stack.append(data)
        self.restore_from_data(data)
        self.update_canvas()

    def restore_from_data(self, data):
        img = Image.open(io.BytesIO(data))
        self.image = img
        self.draw = ImageDraw.Draw(self.image)

    def update_canvas(self):
        self.canvas.delete("all")
        # Просто загружаем изображение на холст
        img = self.image.resize((self.canvas.winfo_width(), self.canvas.winfo_height()), Image.Resampling.LANCZOS)
        self.tk_img = ImageTk.PhotoImage(img)
        self.canvas.create_image(0, 0, anchor=tk.NW, image=self.tk_img)

    def clear(self):
        self.canvas.delete("all")
        self.image = Image.new("RGB", (800, 600), self.bg_color)
        self.draw = ImageDraw.Draw(self.image)
        self.push_undo()

    def save(self):
        filepath = filedialog.asksaveasfilename(defaultextension=".png",
                                                filetypes=[("PNG", "*.png"), ("All", "*.*")])
        if filepath:
            self.image.save(filepath, "PNG")

    def bind_keys(self):
        self.root.bind("<Control-z>", lambda e: self.undo())
        self.root.bind("<Control-y>", lambda e: self.redo())
        self.root.bind("<Control-s>", lambda e: self.save())

if __name__ == "__main__":
    root = tk.Tk()
    app = MousePaint(root)
    root.mainloop()

File: mouse_paint.rs
// mouse_paint.rs - Рисовалка с мышкой на Rust (веб-сервер)
use actix_web::{web, App, HttpResponse, HttpServer, Result};
use base64::{engine::general_purpose, Engine};
use std::fs;

async fn index() -> HttpResponse {
    let html = r#"<!DOCTYPE html>
<html><head><meta charset="UTF-8"><title>🖌️ MousePaint - Rust</title>
<style>body{font-family:sans-serif;background:#2c3e50;display:flex;justify-content:center;align-items:center;min-height:100vh;margin:0;}
.container{background:#ecf0f1;padding:20px;border-radius:16px;box-shadow:0 10px 30px rgba(0,0,0,0.3);}
.toolbar{display:flex;flex-wrap:wrap;gap:6px;margin-bottom:12px;align-items:center;}
.toolbar button{padding:6px 12px;border:none;border-radius:6px;background:#3498db;color:white;cursor:pointer;}
.toolbar button:hover{background:#2980b9;}
.toolbar input[type="color"]{width:40px;height:40px;border:none;cursor:pointer;}
.toolbar input[type="range"]{width:100px;}
canvas{border:2px solid #bdc3c7;border-radius:8px;cursor:crosshair;background:white;}
.status{margin-top:10px;display:flex;justify-content:space-between;font-size:14px;color:#2c3e50;}
</style></head>
<body>
<div class="container">
<h2>🖌️ MousePaint · Rust</h2>
<div class="toolbar">
<button onclick="setTool('brush')">🖌️ Кисть</button>
<button onclick="setTool('eraser')">🧽 Ластик</button>
<input type="color" id="colorPicker" value="#000000">
<label>Толщина: <input type="range" id="sizeSlider" min="1" max="20" value="5"></label>
<button onclick="undo()">↩️ Отменить</button>
<button onclick="redo()">↪️ Повторить</button>
<button onclick="clearCanvas()">🗑️ Очистить</button>
<button onclick="saveImage()">💾 Сохранить</button>
</div>
<canvas id="canvas" width="800" height="600"></canvas>
<div class="status"><span>Инструмент: <span id="toolDisplay">Кисть</span></span><span>Размер: <span id="sizeDisplay">5</span></span></div>
</div>
<script>
const canvas=document.getElementById('canvas');
const ctx=canvas.getContext('2d');
const colorPicker=document.getElementById('colorPicker');
const sizeSlider=document.getElementById('sizeSlider');
let isDrawing=false,lastX=0,lastY=0,tool='brush',color='#000000',size=5;
let undoStack=[],redoStack=[],MAX_UNDO=20;

function init(){ ctx.fillStyle='#ffffff'; ctx.fillRect(0,0,canvas.width,canvas.height); pushUndo(); updateDisplay(); }
function pushUndo(){ undoStack.push(canvas.toDataURL()); if(undoStack.length>MAX_UNDO) undoStack.shift(); redoStack=[]; }
function restoreFromDataURL(data){ let img=new Image(); img.onload=()=>{ ctx.clearRect(0,0,canvas.width,canvas.height); ctx.drawImage(img,0,0); }; img.src=data; }
function undo(){ if(undoStack.length<2)return; redoStack.push(undoStack.pop()); restoreFromDataURL(undoStack[undoStack.length-1]); }
function redo(){ if(redoStack.length===0)return; let data=redoStack.pop(); undoStack.push(data); restoreFromDataURL(data); }
function setTool(t){ tool=t; document.getElementById('toolDisplay').textContent=t==='brush'?'Кисть':'Ластик'; }
function clearCanvas(){ ctx.fillStyle='#ffffff'; ctx.fillRect(0,0,canvas.width,canvas.height); pushUndo(); }
function saveImage(){ let link=document.createElement('a'); link.download='drawing.png'; link.href=canvas.toDataURL('image/png'); link.click();
    let dataURL=canvas.toDataURL('image/png');
    fetch('/save', {method:'POST', body:'image='+encodeURIComponent(dataURL)});
}
function updateDisplay(){ document.getElementById('sizeDisplay').textContent=size; }
function startDrawing(e){ isDrawing=true; let rect=canvas.getBoundingClientRect(); lastX=e.clientX-rect.left; lastY=e.clientY-rect.top; pushUndo(); }
function draw(e){ if(!isDrawing)return; let rect=canvas.getBoundingClientRect(); let x=e.clientX-rect.left; let y=e.clientY-rect.top;
ctx.beginPath(); ctx.moveTo(lastX,lastY); ctx.lineTo(x,y); ctx.strokeStyle=tool==='eraser'?'#ffffff':color; ctx.lineWidth=size; ctx.lineCap='round'; ctx.stroke(); lastX=x; lastY=y; }
function stopDrawing(e){ isDrawing=false; pushUndo(); }
canvas.addEventListener('mousedown',startDrawing);
canvas.addEventListener('mousemove',draw);
canvas.addEventListener('mouseup',stopDrawing);
canvas.addEventListener('mouseleave',stopDrawing);
colorPicker.oninput=(e)=>{ color=e.target.value; };
sizeSlider.oninput=(e)=>{ size=parseInt(e.target.value); updateDisplay(); };
init();
</script>
</body></html>"#;
    HttpResponse::Ok()
        .content_type("text/html; charset=utf-8")
        .body(html)
}

async fn save(web::Form(form): web::Form<serde_json::Value>) -> HttpResponse {
    if let Some(image_data) = form.get("image").and_then(|v| v.as_str()) {
        if let Some(encoded) = image_data.strip_prefix("data:image/png;base64,") {
            if let Ok(decoded) = general_purpose::STANDARD.decode(encoded) {
                if let Ok(()) = fs::write("drawing.png", decoded) {
                    return HttpResponse::Ok().body("ok");
                }
            }
        }
    }
    HttpResponse::BadRequest().body("error")
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| {
        App::new()
            .route("/", web::get().to(index))
            .route("/save", web::post().to(save))
    })
    .bind("127.0.0.1:8000")?
    .run()
    .await
}
